src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs

```diff
@@ -52,6 +51,36 @@ public async Task Invoke(HttpContext httpContext)
             }
         }
 
+        private async Task<AuthenticateResult> AuthenticateAsync(HttpContext httpContext, DownstreamRoute route)
+        {
+            if (!string.IsNullOrWhiteSpace(route.AuthenticationOptions.AuthenticationProviderKey))
+            {
+                return await httpContext.AuthenticateAsync(route.AuthenticationOptions.AuthenticationProviderKey);
+            }
+
+            AuthenticateResult result = null;
+
+            if (route.AuthenticationOptions.AuthenticationProviderKeys is not null)
+            {
+                IEnumerable<string> authenticationProviderKeys =
+                    route
+                    .AuthenticationOptions
+                    .AuthenticationProviderKeys
+                    .Where(apk => !string.IsNullOrWhiteSpace(apk));
+
+                foreach (var authenticationProviderKey in authenticationProviderKeys)
+                {
+                    result = await httpContext.AuthenticateAsync(authenticationProviderKey);
+                    if (result.Succeeded)
+                    {
+                        return result;
+                    }
+                }
+            }
+
+            return result ?? AuthenticateResult.NoResult();
```

Comment:
What do you think of this example? I tend to try to reduce the nesting, but that's cosmetics...

```c#
var result = AuthenticateResult.NoResult();

if (route.AuthenticationOptions.AuthenticationProviderKeys.Length == 0)
{
    return result;
}

var authenticationProviderKeys =
    route
        .AuthenticationOptions
        .AuthenticationProviderKeys
        .Where(apk => !string.IsNullOrWhiteSpace(apk));

foreach (var authenticationProviderKey in authenticationProviderKeys)
{
    result = await httpContext.AuthenticateAsync(authenticationProviderKey);
    if (result.Succeeded)
    {
        return result;
    }
}

return result;
```

Reply:
Resolved.